nautobot_ssot/contrib.py

```diff
@@ -0,0 +1,354 @@
+
+            # As the default case, just set the attribute directly
+            setattr(obj, field, value)
+
+        # Set foreign keys
+        cls._lookup_and_set_foreign_keys(foreign_keys, obj)
+
+        # Save the object to the database
+        try:
+            obj.validated_save()
+        except ValidationError as error:
+            raise ValidationError(f"Parameters: {parameters}") from error
+
+        # Set many-to-many fields after saving
+        cls._set_many_to_many_fields(many_to_many_fields, obj)
+
+    @classmethod
+    def _set_many_to_many_fields(cls, many_to_many_fields, obj):
+        """
+        Given a dictionary, set many-to-many relationships on an object.
+
+        This will always use `set`, thereby replacing any elements that are already part of the relationship.
+
+        Example dictionary:
+        {
+            "relationship_field_name": [<Object 1>, <Object 2>],
+            ...
+        }
+        """
+        for field_name, related_objects in many_to_many_fields.items():
+            many_to_many_field = getattr(obj, field_name)
+            many_to_many_field.set(related_objects)
+
+    @classmethod
+    def _lookup_and_set_foreign_keys(cls, foreign_keys, obj):
+        """
+        Given a list of foreign keys as dictionaries, look up and set foreign keys on an object.
+
+        Dictionary should be in the form of:
+        [
+          {"field_1": "value_1", "field_2": "value_2"},
+          ...
+        ]
+        where each item in the list corresponds to the parameters needed to uniquely identify a foreign key object.
+        """
+        for field_name, related_model_dict in foreign_keys.items():
+            related_model = related_model_dict.pop("_model_class")
+            # Generic foreign keys will not have this dictionary field. As such, we need to retrieve the appropriate
+            # model class through other means.
+            if not related_model:
+                try:
+                    app_label = related_model_dict.pop("app_label")
+                    model = related_model_dict.pop("model")
+                except KeyError as error:
+                    raise ValueError(
+                        f"Missing annotation for '{field_name}__app_label' or '{field_name}__model - this is required"
+                        f"for generic foreign keys."
+                    ) from error
+                related_model = ContentType.objects.get(app_label=app_label, model=model).model_class()
+            # Set the foreign key to 'None' when none of the fields are set to anything
+            if not any(related_model_dict.values()):
+                setattr(obj, field_name, None)
+                continue
+            try:
+                related_object = related_model.objects.get(**related_model_dict)
+            except related_model.DoesNotExist as error:
+                raise ValueError(f"Couldn't find {field_name} instance with: {related_model_dict}.") from error
```

Comment:
Maybe also catch `MultipleObjectsReturned` and reraise?

Reply:
Done